File: backend/models/wallpaper.go
package models

import (
	"encoding/json"
	"strconv"
)

// FlexInt accepts both JSON number and string when unmarshaling
type FlexInt int

func (f *FlexInt) UnmarshalJSON(data []byte) error {
	if len(data) == 0 {
		return nil
	}
	// Try number first
	if data[0] != '"' {
		var i int
		if err := json.Unmarshal(data, &i); err == nil {
			*f = FlexInt(i)
			return nil
		}
	}
	// Try string
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}
	n, err := strconv.Atoi(s)
	if err != nil {
		return err
	}
	*f = FlexInt(n)
	return nil
}

// WallhavenResponse is the top-level response from /api/v1/search
type WallhavenResponse struct {
	Data []WallhavenImage `json:"data"`
	Meta struct {
		CurrentPage FlexInt `json:"current_page"`
		LastPage    FlexInt `json:"last_page"`
		PerPage     FlexInt `json:"per_page"`
		Total       FlexInt `json:"total"`
	} `json:"meta"`
}

// WallhavenImage represents a single wallpaper from the Wallhaven API
type WallhavenImage struct {
	ID         string   `json:"id"`
	Path       string   `json:"path"`
	Resolution string   `json:"resolution"`
	FileSize   int64    `json:"file_size"`
	DimensionX int      `json:"dimension_x"`
	DimensionY int      `json:"dimension_y"`
	Category   string   `json:"category"`
	Colors     []string `json:"colors"`
	Thumbs     struct {
		Large    string `json:"large"`
		Original string `json:"original"`
		Small    string `json:"small"`
	} `json:"thumbs"`
}

// WallpaperInsert maps directly to the Supabase wallpapers table columns
type WallpaperInsert struct {
	WallhavenID  string `json:"wallhaven_id"`
	URLFull      string `json:"url_full"`
	URLThumb     string `json:"url_thumb"`
	Resolution   string `json:"resolution"`
	Width        int    `json:"width"`
	Height       int    `json:"height"`
	FileSize     int64  `json:"file_size"`
	PrimaryColor string `json:"primary_color"`
	Category     string `json:"category"`
	SourceQuery  string `json:"source_query"`
}

// HealthResponse is the response for the health check endpoint
type HealthResponse struct {
	Status string `json:"status" example:"ok"`
}

// SyncResponse is the response for the manual sync trigger endpoint
type SyncResponse struct {
	Status string `json:"status" example:"sync triggered"`
}

// ErrorResponse is a generic error response
type ErrorResponse struct {
	Error string `json:"error" example:"something went wrong"`
}

File: app/lib/models/wallpaper.dart
class Wallpaper {
  final String id;
  final String wallhavenId;
  final String urlFull;
  final String urlThumb;
  final String resolution;
  final int width;
  final int height;
  final int fileSize;
  final String primaryColor;
  final String category;
  final String sourceQuery;
  final DateTime createdAt;

  Wallpaper({
    required this.id,
    required this.wallhavenId,
    required this.urlFull,
    required this.urlThumb,
    required this.resolution,
    required this.width,
    required this.height,
    required this.fileSize,
    required this.primaryColor,
    required this.category,
    required this.sourceQuery,
    required this.createdAt,
  });

  double get aspectRatio => height == 0 ? 1.0 : width / height;

  String get formattedFileSize {
    if (fileSize < 1024 * 1024) {
      return '${(fileSize / 1024).toStringAsFixed(1)} KB';
    }
    return '${(fileSize / (1024 * 1024)).toStringAsFixed(1)} MB';
  }

  factory Wallpaper.fromMap(Map<String, dynamic> map) {
    final id = map['id'] as String? ?? '';
    final wallhavenId = map['wallhaven_id'] as String? ?? '';
    final urlFull = map['url_full'] as String? ?? '';
    if (id.isEmpty && wallhavenId.isEmpty) {
      throw FormatException('Wallpaper missing both id and wallhaven_id');
    }
    return Wallpaper(
      id: id,
      wallhavenId: wallhavenId,
      urlFull: urlFull,
      urlThumb: (map['url_thumb'] as String?) ?? '',
      resolution: (map['resolution'] as String?) ?? '',
      width: (map['width'] as int?) ?? 0,
      height: (map['height'] as int?) ?? 0,
      fileSize: (map['file_size'] as int?) ?? 0,
      primaryColor: (map['primary_color'] as String?) ?? '#000000',
      category: (map['category'] as String?) ?? 'general',
      sourceQuery: (map['source_query'] as String?) ?? '',
      createdAt:
          DateTime.tryParse(map['created_at'] as String? ?? '') ??
          DateTime.now(),
    );
  }

  factory Wallpaper.fromWallhavenMap(Map<String, dynamic> map) {
    final colors = map['colors'] as List<dynamic>?;
    final thumbs = map['thumbs'] as Map<String, dynamic>?;
    return Wallpaper(
      id: 'wh-${map['id'] ?? ''}',
      wallhavenId: map['id'] ?? '',
      urlFull: map['path'] ?? '',
      urlThumb:
          (thumbs?['small'] as String?) ??
          (thumbs?['original'] as String?) ??
          '',
      resolution: map['resolution'] ?? '',
      width: map['dimension_x'] ?? 0,
      height: map['dimension_y'] ?? 0,
      fileSize: map['file_size'] ?? 0,
      primaryColor: (colors != null && colors.isNotEmpty)
          ? colors[0] as String
          : '#000000',
      category: map['category'] ?? 'general',
      sourceQuery: '',
      createdAt: DateTime.tryParse(map['created_at'] ?? '') ?? DateTime.now(),
    );
  }
}
